forte/sci/fci_mo.cc

```diff
@@ -2703,4 +2706,777 @@ void FCI_MO::print_occupation_strings_perirrep(
+        }
+    }
+
+    std::vector<std::vector<std::tuple<size_t, short, short>>>& ab_list = op->ab_list_;
+    for (auto& coupled_dets : ab_list) {
+        for (size_t a = 0, max_a = coupled_dets.size(); a < max_a; ++a) {
+            auto& detJ = coupled_dets[a];
+
+            const size_t& J = std::get<0>(detJ);
+            const size_t& p = std::abs(std::get<1>(detJ)) - 1;
+            const size_t& q = std::get<2>(detJ);
+            const double& sign_pq = std::get<1>(detJ) > 0.0 ? 1.0 : -1.0;
+
+            for (size_t b = a + 1, max_b = coupled_dets.size(); b < max_b; ++b) {
+                auto& detI = coupled_dets[b];
+
+                const size_t& r = std::abs(std::get<1>(detI)) - 1;
+                const size_t& s = std::get<2>(detI);
+                const double& sign_rs = std::get<1>(detI) > 0.0 ? 1.0 : -1.0;
+                const size_t& I = std::get<0>(detI);
+
+                double sign = sign_pq * sign_rs;
+
+                cc2ab_data[J * na5 + I * na4 + p * na3 + q * na2 + r * na1 + s] = sign;
+
+                cc2ab_data[I * na5 + J * na4 + r * na3 + s * na2 + p * na1 + q] = sign;
+            }
+        }
+    }
+
+    return out;
+}
+
+std::vector<ambit::Tensor> FCI_MO::coupling_coeffcients_3() {
+    size_t ndets = determinant_.size();
+
+    size_t na1 = nactv_;
+    size_t na2 = na1 * na1;
+    size_t na3 = na2 * na1;
+    size_t na4 = na3 * na1;
+    size_t na5 = na4 * na1;
+    size_t na6 = na5 * na1;
+    size_t na7 = na6 * ndets;
+
+    // stored <I| p^+ q^+ r^+ u t s |J> as I->0, J->1, p->2, q->3, r->4, s->5, t->6, u->7
+    std::vector<ambit::Tensor> out(4);
+    std::vector<size_t> dims{ndets, ndets, na1, na1, na1, na1, na1, na1};
+    out[0] = ambit::Tensor::build(ambit::CoreTensor, "CC3aaa", dims);
+    out[1] = ambit::Tensor::build(ambit::CoreTensor, "CC3aab", dims);
+    out[2] = ambit::Tensor::build(ambit::CoreTensor, "CC3abb", dims);
+    out[3] = ambit::Tensor::build(ambit::CoreTensor, "CC3bbb", dims);
+    auto& cc3aaa_data = out[0].data();
+    auto& cc3aab_data = out[1].data();
+    auto& cc3abb_data = out[2].data();
+    auto& cc3bbb_data = out[3].data();
+
+    // compute coupling lists
+    const auto& dets = det_hash_vec_.wfn_hash();
+    auto op = std::make_shared<DeterminantSubstitutionLists>(fci_ints_);
+    op->set_quiet_mode(not print_);
+    op->build_strings(det_hash_vec_);
+    op->three_s_lists(det_hash_vec_);
+
+    // diagonal
+    for (size_t I = 0; I < ndets; ++I) {
+        size_t idx_I = I * na7 + I * na6;
+
+        std::vector<int> aocc = dets[I].get_alfa_occ(na1);
+        std::vector<int> bocc = dets[I].get_beta_occ(na1);
+
+        int naocc = aocc.size();
+        int nbocc = bocc.size();
+
+        // aaa
+        for (int p = 0; p < naocc; ++p) {
+            int pp = aocc[p];
+            for (int q = p + 1; q < naocc; ++q) {
+                int qq = aocc[q];
+                for (int r = q + 1; r < naocc; ++r) {
+                    int rr = aocc[r];
```

Comment:
Any ideas why we're not getting code coverage here?

Reply:
Not sure why, but the test `dsrg-mrpt2-gradient-1` uses CAS(4,4) and should have tested this.